### ISSUE TITLE
Address Sanitizer error with jemalloc

### ISSUE THREAD

**Issue by res0nance** [jemalloc/jemalloc#2543]
We encountered the following ASAN error. We can't reproduce this reliably so we can't provide a reproducible example.

I hope this stacktrace will mean something to a reader.

jemalloc version: 5.3.0

```
AddressSanitizer:DEADLYSIGNAL
=================================================================
==369==ERROR: AddressSanitizer: SEGV on unknown address 0x007f9fd9d1a0 (pc 0x557f3ee485c5 bp 0x000000000000 sp 0x7ffe82eb8c80 T0)
==369==The signal is caused by a WRITE memory access.
    #0 0x557f3ee485c5 in edata_list_inactive_remove /project/third_party/jemalloc/include/jemalloc/internal/edata.h:254:1
    #1 0x557f3ee485c5 in je_eset_remove /project/third_party/jemalloc/src/eset.c:145:2
    #2 0x557f3ee49c92 in je_ecache_evict /project/third_party/jemalloc/src/extent.c:182:3
    #3 0x557f3ee5435c in pac_stash_decayed /project/third_party/jemalloc/src/pac.c:351:20
    #4 0x557f3ee5435c in pac_decay_to_limit /project/third_party/jemalloc/src/pac.c:449:18
    #5 0x557f3ee546e4 in je_pac_maybe_decay_purge /project/third_party/jemalloc/src/pac.c
    #6 0x557f3ee238ed in arena_decay_impl /project/third_party/jemalloc/src/arena.c:439:24
    #7 0x557f3ee1f397 in arena_decay_dirty /project/third_party/jemalloc/src/arena.c:459:9
    #8 0x557f3ee1f397 in je_arena_decay /project/third_party/jemalloc/src/arena.c:485:6
    #9 0x557f3ee2070c in arena_decay_ticks /project/third_party/jemalloc/include/jemalloc/internal/arena_inlines_b.h:135:3
    #10 0x557f3ee2070c in arena_decay_tick /project/third_party/jemalloc/include/jemalloc/internal/arena_inlines_b.h:141:2
    #11 0x557f3ee2070c in je_arena_cache_bin_fill_small /project/third_party/jemalloc/src/arena.c:1052:2
    #12 0x557f3ee5cf96 in je_tcache_alloc_small_hard /project/third_party/jemalloc/src/tcache.c:238:2
    #13 0x557f3ee11733 in tcache_alloc_small /project/third_party/jemalloc/include/jemalloc/internal/tcache_inlines.h:68:9
    #14 0x557f3ee11733 in arena_malloc /project/third_party/jemalloc/include/jemalloc/internal/arena_inlines_b.h:151:11
    #15 0x557f3ee11733 in iallocztm /project/third_party/jemalloc/include/jemalloc/internal/jemalloc_internal_inlines_c.h:55:8
    #16 0x557f3ee11733 in imalloc_no_sample /project/third_party/jemalloc/src/jemalloc.c:2398:9
    #17 0x557f3ee11733 in imalloc_body /project/third_party/jemalloc/src/jemalloc.c:2573:16
    #18 0x557f3ee11733 in imalloc /project/third_party/jemalloc/src/jemalloc.c:2687:10
    #19 0x557f3ee11733 in je_malloc_default /project/third_party/jemalloc/src/jemalloc.c:2722:2
```

**Comment by interwq** [comment]
This path is typically a heap corruption from the application, e.g. from double frees or invalid pointers.

Also for ASAN build you should not link against jemalloc. ASAN has its own allocator which cannot be replaced. If you use jemalloc and ASAN at the same time, one of them is not working properly.

**Comment by hunjmes** [comment]
We (a different we!) are encountering ~~the same~~ **a similar** failing call stack, same SEGV, **without** ASAN, intermittently in our code.

I agree that it is likely that we have a double free, or have somehow else corrupted jemalloc's data structures, before we reach this SEGV. But note that our SEGV occurs deep inside `jemalloc`'s extent management code:
```
#25 <signal handler called>
#26 0x00007fba7b25cf1e in edata_list_inactive_remove (item=0x7cba5f616580, list=<optimized out>) at include/jemalloc/internal/edata.h:254
#27 je_eset_remove (eset=0x7cba5f603a98, edata=edata@entry=0x7cba5f616580) at src/eset.c:145
#28 0x00007fba7b26090b in extent_activate_locked (edata=0x7cba5f616580, eset=<optimized out>, ecache=0x7cba5f603a28, pac=0x7cba5f6039f0, tsdn=0x7cba64df6838) at src/extent.c:280
#29 extent_recycle_extract (ehooks=0x7cba5f6000c0, guarded=<optimized out>, alignment=4096, size=4096, expand_edata=0x0, ecache=0x7cba5f603a28, pac=0x7cba5f6039f0, tsdn=0x7cba64df6838)
    at src/extent.c:444
#30 extent_recycle (tsdn=tsdn@entry=0x7cba64df6838, pac=pac@entry=0x7cba5f6039f0, ehooks=ehooks@entry=0x7cba5f6000c0, ecache=ecache@entry=0x7cba5f603a28, expand_edata=expand_edata@entry=0x0, 
    size=size@entry=4096, alignment=<optimized out>, zero=<optimized out>, commit=<optimized out>, growing_retained=<optimized out>, guarded=<optimized out>) at src/extent.c:606
#31 0x00007fba7b260c69 in je_ecache_alloc (tsdn=tsdn@entry=0x7cba64df6838, pac=pac@entry=0x7cba5f6039f0, ehooks=ehooks@entry=0x7cba5f6000c0, ecache=ecache@entry=0x7cba5f603a28, 
    expand_edata=expand_edata@entry=0x0, size=size@entry=4096, alignment=4096, zero=false, guarded=false) at src/extent.c:87
#32 0x00007fba7b26c1ce in pac_alloc_real (tsdn=tsdn@entry=0x7cba64df6838, pac=pac@entry=0x7cba5f6039f0, ehooks=ehooks@entry=0x7cba5f6000c0, size=size@entry=4096, alignment=alignment@entry=4096, 
    zero=zero@entry=false, guarded=false) at src/pac.c:116
#33 0x00007fba7b26c36c in pac_alloc_impl (tsdn=0x7cba64df6838, self=0x7cba5f6039f0, size=4096, alignment=4096, zero=<optimized out>, guarded=<optimized out>, frequent_reuse=true, 
    deferred_work_generated=0x7cba64df22df) at src/pac.c:178
#34 0x00007fba7b26b1e1 in pai_alloc (deferred_work_generated=0x7cba64df22df, frequent_reuse=true, guarded=<optimized out>, zero=false, alignment=<optimized out>, size=4096, self=0x7cba5f6039f0, 
    tsdn=0x7cba64df6838) at include/jemalloc/internal/pai.h:43
#35 je_pa_alloc (tsdn=tsdn@entry=0x7cba64df6838, shard=shard@entry=0x7cba5f6039d8, size=4096, alignment=alignment@entry=4096, slab=slab@entry=true, szind=szind@entry=4, zero=false, guarded=false, 
    deferred_work_generated=0x7cba64df22df) at src/pa.c:139
#36 0x00007fba7b20d8dd in arena_slab_alloc (tsdn=tsdn@entry=0x7cba64df6838, arena=arena@entry=0x7cba5f601040, binind=binind@entry=4, binshard=binshard@entry=0, 
    bin_info=bin_info@entry=0x7fba7b4bde20 <je_bin_infos+160>) at src/arena.c:842
#37 0x00007fba7b20f195 in je_arena_cache_bin_fill_small (tsdn=<optimized out>, arena=<optimized out>, cache_bin=<optimized out>, cache_bin_info=<optimized out>, binind=<optimized out>, 
    nfill=<optimized out>) at src/arena.c:1037
#38 0x00007fba7b282d89 in je_tcache_alloc_small_hard (tsdn=tsdn@entry=0x7cba64df6838, arena=arena@entry=0x7cba5f601040, tcache=tcache@entry=0x7cba64df6b90, 
    cache_bin=cache_bin@entry=0x7cba64df6bf8, binind=binind@entry=4, tcache_success=tcache_success@entry=0x7cba64df2450) at src/tcache.c:238
#39 0x00007fba7b1fc85f in tcache_alloc_small (slow_path=<optimized out>, zero=false, binind=4, size=<optimized out>, tcache=0x7cba64df6b90, arena=0x7cba5f601040, tsd=0x7cba64df6838)
    at include/jemalloc/internal/tcache_inlines.h:68
#40 arena_malloc (slow_path=<optimized out>, tcache=0x7cba64df6b90, zero=false, ind=4, size=<optimized out>, arena=0x0, tsdn=0x7cba64df6838) at include/jemalloc/internal/arena_inlines_b.h:151
#41 iallocztm (slow_path=<optimized out>, arena=0x0, is_internal=false, tcache=0x7cba64df6b90, zero=false, ind=4, size=<optimized out>, tsdn=0x7cba64df6838)
    at include/jemalloc/internal/jemalloc_internal_inlines_c.h:55
#42 imalloc_no_sample (ind=4, usize=64, size=<optimized out>, tsd=0x7cba64df6838, dopts=<synthetic pointer>, sopts=<synthetic pointer>) at src/jemalloc.c:2411
#43 imalloc_body (tsd=0x7cba64df6838, dopts=<synthetic pointer>, sopts=<synthetic pointer>) at src/jemalloc.c:2586
#44 imalloc (dopts=<synthetic pointer>, sopts=<synthetic pointer>) at src/jemalloc.c:2703
#45 je_malloc_default (size=<optimized out>) at src/jemalloc.c:2738
#46 0x00007fba7b28a2f9 in fallback_impl<false> (size=52) at src/jemalloc_cpp.cpp:104
#47 0x00007fba7b28a3ce in imalloc_fastpath (fallback_alloc=0x7fba7b28a2f0 <fallback_impl<false>(std::size_t)>, size=<optimized out>) at include/jemalloc/internal/jemalloc_internal_inlines_c.h:266
#48 0x00000000005b665e in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*> (this=this@entry=0x7cba64df2620, 
    __beg=0x7fba6db5cb40 "<some 51 character string>", __end=<optimized out>)
    at .../include/c++/11.4.0/bits/basic_string.tcc:219
#49 0x00000000005c5c9c in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct_aux<char*> (__end=<optimized out>, __beg=<optimized out>, 
    this=0x7cba64df2620) at .../include/c++/11.4.0/bits/basic_string.h:255
#50 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::_M_construct<char*> (__end=<optimized out>, __beg=<optimized out>, this=0x7cba64df2620)
    at .../include/c++/11.4.0/bits/basic_string.h:274
#51 std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string (__str=..., this=0x7cba64df2620)
    at .../include/c++/11.4.0/bits/basic_string.h:459
#52 std::operator+<char, std::char_traits<char>, std::allocator<char> > (__lhs=..., __rhs=0x7fba7c767246 "-")
    at .../include/c++/11.4.0/bits/basic_string.h:6135
```

A few other notes: first, we have tried to repro this with `jemalloc` configured for DEBUG mode, several times, and have not able to repro under DEBUG. It is possible this is just because this crash occurs so rarely in our tests -- maybe we just missed the necessary conditions, during our DEBUG runs. 

Second, this crash occurs when our program starts up. Now, we do a lot of things during startup, any of which could possibly corrupt memory -- but then, so does `jemalloc`. 

Last, for what it's worth, this SEGV always occurs, when it does occur, in our code, with exactly the same call stack pasted above: under a small allocation, under `std::string`'s copy constructor, while our code is starting up. It is possible that this small allocation just happens to be the first allocation after `jemalloc` got corrupted; or, there might be a race condition inside `jemalloc`, when populating extent caches.

**Comment by hunjmes** [comment]
-- I guess what I am saying is: suppose we did double-free a chunk of memory, via `jemalloc`. How would that cause `jemalloc` to SEGV when evicting an "eset", while trying to allocate a new arena, of size 4 KB?

**Comment by hunjmes** [comment]
Immediate cause of SEGV:
```
(gdb) frame 27
#27 je_eset_remove (eset=0x7cba5f603a98, edata=edata@entry=0x7cba5f616580) at src/eset.c:145
145	in src/eset.c
(gdb) p eset->lru
$16 = {head = {qlh_first = 0x0}}
(gdb) p edata
$17 = (edata_t *) 0x7cba5f616580
```
Dying inside call:
```
edata_list_inactive_remove(&eset->lru, edata);
```
So, initial glance at the code suggests that the problem is that `eset->lru` is unexpectedly an empty list. Caller apparently tries to remove `edata` from `eset->lru`, and SEGVs when it can't find `edata` in that list, since the list is unexpectedly empty.

**Comment by hunjmes** [comment]
The failing code all lives under function `extent_recycle()`, which first does:
```
malloc_mutex_lock(tsdn, &ecache->mtx);
```
Looking at the mutex:
```
(gdb) p ecache->mtx
$53 = {{{prof_data = {tot_wait_time = {ns = 0}, max_wait_time = {ns = 0}, n_wait_times = 0, n_spin_acquired = 0, max_n_thds = 0, n_waiting_thds = {repr = 1}, n_owner_switches = 5, 
        prev_owner = 0x7cba64df6838, n_lock_ops = 25}, lock = {__data = {__lock = 2, __count = 0, __owner = 123287, __nusers = 1, __kind = 512, __spins = 0, __elision = 0, __list = {__prev = 0x0, 
            __next = 0x0}}, __size = "\002\000\000\000\000\000\000\000\227\341\001\000\001\000\000\000\000\002", '\000' <repeats 21 times>, __align = 2}, locked = {repr = false}}, witness = {
      name = 0x0, rank = WITNESS_RANK_OMIT, comp = 0x0, opaque = 0x0, link = {qre_next = 0x100000000, qre_prev = 0x5}}, lock_order = malloc_mutex_rank_exclusive}}
```
The `__owner` is correct: the SEGV-ing thread is, indeed, `LWP 123287`. So we own the mutex.

Since this SEGV occurs when `jemalloc` tries to access its own data structures, while it holds its own mutex, I think there's a decent probability that the bug occurs inside the `jemalloc` library itself. 

Basically: we acquire `ecache->mtx`, then try to recycle an extent from `ecache->eset` (since `guarded` is `false`). We find an `edata`, via a call to `eset_fit()`. However, when we try to remove this `edata` from `ecache->eset.lru`, we SEGV, since the `lru` list is empty.

**Comment by hunjmes** [comment]
Continuing investigation: the `eset` is completely empty (since the bitmap is all zeroes):
```
(gdb) p *eset
$62 = {bitmap = {0, 0, 0, 0}, bins = {{heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, addr = 137139903795200}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, 
        addr = 0}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, addr = 137139903877120}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, 
        addr = 137139903873024}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, addr = 137139903799296}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, 
        addr = 137139903819776}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, addr = 137139903791104}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, 
        addr = 0}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, addr = 0}}, {heap = {ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, addr = 137139903799296}}, {heap = {
        ph = {root = 0x0, auxcount = 0}}, heap_min = {sn = 0, addr = 0}} <repeats 190 times>}, bin_stats = {{nextents = {repr = 0}, nbytes = {repr = 0}} <repeats 200 times>}, lru = {head = {
      qlh_first = 0x0}}, npages = {repr = 1}, state = extent_state_dirty}
```
And the `eset_s` class is explicitly **not** thread safe -- must be accessed while holding mutex -- so it seems like a "double free" couldn't cause problems. Because, each of the "frees" will hold the mutex while it tries to recycle extents, so whether it finds an extent to recycle, it still won't leave the structures in an inconsistent state.

Somehow, function `eset_fit()` returned a non-NULL `edata`, from an empty `eset`??

**Comment by hunjmes** [comment]
The `edata` is a 4KB slab, so its `e_bits` seem wrong to me:
```
(gdb) p /x *edata
$145 = {e_bits = 0xe823011, e_addr = 0x7cba5f401000, {e_size_esn = 0x1000, e_bsize = 0x1000}, e_ps = 0x0, e_sn = 0x0, {ql_link_active = {qre_next = 0x0, qre_prev = 0x0}, {heap_link = {link = {
          prev = 0x0, next = 0x0, lchild = 0x0}}, avail_link = {link = {prev = 0x0, next = 0x0, lchild = 0x0}}}}, {ql_link_inactive = {qre_next = 0x0, qre_prev = 0x0}, e_slab_data = {bitmap = {
        0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}}, e_prof_info = {e_prof_alloc_time = {ns = 0x0}, e_prof_alloc_size = 0x0, e_prof_tctx = {repr = 0x0}, e_prof_recent_alloc = {repr = 0x0}}}}
```

If I read the `edata.h` comments correctly, this means `szind` = 232? But this is a 4 KB slab, so shouldn't this be `0` ?
